File: p4+/main.go
package main

import (
	"bufio"
	"cmp"
	"fmt"
	"io"
	"math/big"
	"os"
	"sort"
	"strconv"
	"unsafe"
)

func solve(t, s int, v []int) *big.Int {
	count := big.NewInt(0)

	sort.Ints(v)

	s0 := int64(v[0]) * int64(t) // max: 1e6 * 1e6 = 1e12
	s0_i := 0
	s0_div_s := s0 / int64(s) // max: 1e12
	s0_mod_s := s0 % int64(s) // max: 1e12

	var sum_div_s, si_count int64

	nodes := make([]Node[int64], len(v))
	new_node := func() *Node[int64] {
		node := &nodes[0]
		nodes = nodes[1:]
		return node
	}

	var tree *Node[int64]
	tree = tree.Insert(s0_mod_s, new_node)

	for i := 1; i < len(v); i++ {
		si := int64(v[i]) * int64(t) // max: 1e6 * 1e6 = 1e12

		if si != s0 {
			sum_div_s += s0_div_s * int64(i-s0_i)        // max: 1e12 * 1e6 = 1e18
			si_div_s := si / int64(s)                    // max: 1e12
			si_count = (si_div_s-1)*int64(i) - sum_div_s // must be > 0, max: (1e12 * 1e6) - 1e18 = 1e18
			s0 = si
			s0_i = i
			s0_div_s = si_div_s
		}

		si_mod_s := si % int64(s)                // max: 1e12
		si_last := int64(tree.FindIdx(si_mod_s)) // max: 1e6
		tree = tree.Insert(si_mod_s, new_node)
		s0_mod_s = si_mod_s

		// count += si_count + si_last // max: (1e18 + 1e6)+... !!! may be int64 overflow !!!
		count.Add(count, big.NewInt(si_count+si_last))
	}

	return count
}

// ----------------------------------------------------------------------------

type Node[K cmp.Ordered] struct {
	key   K
	left  *Node[K]
	right *Node[K]
	size  int
	hght  int
}

func (n *Node[K]) Key() K {
	if n == nil {
		return *new(K)
	}
	return n.key
}

func (n *Node[K]) Size() int {
	if n == nil {
		return 0
	}
	return n.size
}

func (n *Node[K]) height() int {
	if n == nil {
		return 0
	}
	return n.hght
}

// Возвращает индекс (начиная с 0) первого не меньшиго по ключу узла.
// Если такого ключа не существует, возвращает n.Size()
func (n *Node[K]) FindIdx(key K) int {
	if n == nil {
		return 0
	}
	if key <= n.key {
		return n.left.FindIdx(key)
	}
	return n.right.FindIdx(key) + n.left.Size() + 1
}

// (!) Как ни странно, это медленние рекурсии
// // Возвращает индекс (начиная с 0) первого не меньшиго по ключу узла.
// // Если такого ключа не существует, возвращает n.Size()
// func (n *Node[K]) FindIdx(key K) int {
// 	idx := 0
// 	for n != nil {
// 		if key <= n.key {
// 			n = n.left
// 			continue
// 		}
// 		idx += n.left.Size() + 1
// 		n = n.right
// 	}
// 	return idx
// }

func (n *Node[K]) Insert(key K, new_node func() *Node[K]) *Node[K] {
	if n == nil {
		n := new_node()
		*n = Node[K]{
			key:  key,
			size: 1,
			hght: 1,
		}
		return n
	}

	if key <= n.key {
		new_left := n.left.Insert(key, new_node)
		n.left = new_left
		return n.repair()
	}

	new_right := n.right.Insert(key, new_node)
	n.right = new_right
	return n.repair()
}

func (n *Node[K]) update() {
	n.size = n.left.Size() + n.right.Size() + 1
	n.hght = max(n.left.height(), n.right.height()) + 1
}

func (n *Node[K]) repair() *Node[K] {
	d := n.left.height() - n.right.height()
	if d < -1 {
		return n.leftRotate()
	}
	if d > 1 {
		return n.rightRotate()
	}
	n.update()
	return n
}

func (n *Node[K]) leftRotate() *Node[K] {
	al := n
	bt := al.right

	if bt.right.height()-bt.left.height() > 0 {
		al.right = bt.left
		al.update()
		bt.left = al
		bt.update()
		return bt
	}

	ga := bt.left
	al.right = ga.left
	al.update()
	bt.left = ga.right
	bt.update()
	ga.left = al
	ga.right = bt
	ga.update()
	return ga
}

func (n *Node[K]) rightRotate() *Node[K] {
	al := n
	bt := al.left

	if bt.left.height()-bt.right.height() > 0 {
		al.left = bt.right
		al.update()
		bt.right = al
		bt.update()
		return bt
	}

	ga := bt.right
	al.left = ga.right
	al.update()
	bt.right = ga.left
	bt.update()
	ga.right = al
	ga.left = bt
	ga.update()
	return ga
}

// ----------------------------------------------------------------------------

func run(in io.Reader, out io.Writer) {
	sc := bufio.NewScanner(in)
	sc.Split(bufio.ScanWords)
	bw := bufio.NewWriter(out)
	defer bw.Flush()

	n, t, s, err := scanThreeInt(sc)
	if err != nil {
		panic(err)
	}

	v := make([]int, n)
	if err := scanInts(sc, v); err != nil {
		panic(err)
	}

	res := solve(t, s, v)

	fmt.Fprintf(bw, "%d\n", res)
}

// ----------------------------------------------------------------------------

var _, debugEnable = os.LookupEnv("DEBUG")

func main() {
	_ = debugEnable
	run(os.Stdin, os.Stdout)
}

// ----------------------------------------------------------------------------

func unsafeString(b []byte) string {
	return *(*string)(unsafe.Pointer(&b))
}

func scanInt(sc *bufio.Scanner) (int, error) {
	sc.Scan()
	return strconv.Atoi(unsafeString(sc.Bytes()))
}

func scanTwoInt(sc *bufio.Scanner) (v1, v2 int, err error) {
	v1, err = scanInt(sc)
	if err == nil {
		v2, err = scanInt(sc)
	}
	return v1, v2, err
}

func scanThreeInt(sc *bufio.Scanner) (v1, v2, v3 int, err error) {
	v1, err = scanInt(sc)
	if err == nil {
		v2, err = scanInt(sc)
	}
	if err == nil {
		v3, err = scanInt(sc)
	}
	return v1, v2, v3, err
}

func scanFourInt(sc *bufio.Scanner) (v1, v2, v3, v4 int, err error) {
	v1, err = scanInt(sc)
	if err == nil {
		v2, err = scanInt(sc)
	}
	if err == nil {
		v3, err = scanInt(sc)
	}
	if err == nil {
		v4, err = scanInt(sc)
	}
	return v1, v2, v3, v4, err
}

func scanInts(sc *bufio.Scanner, a []int) error {
	for i := range a {
		v, err := scanInt(sc)
		if err != nil {
			return err
		}
		a[i] = v
	}
	return nil
}

type Int interface {
	~int | ~int64 | ~int32 | ~int16 | ~int8
}

type writeOpts struct {
	sep byte
	end byte
}

func writeInt[I Int](bw *bufio.Writer, v I, opts writeOpts) error {
	var buf [32]byte

	_, err := bw.Write(strconv.AppendInt(buf[:0], int64(v), 10))

	if err == nil && opts.end != 0 {
		bw.WriteByte(opts.end)
	}

	return err
}

func writeInts[I Int](bw *bufio.Writer, a []I, opts writeOpts) error {
	var err error

	if len(a) != 0 {
		var buf [32]byte

		if opts.sep == 0 {
			opts.sep = ' '
		}

		_, err = bw.Write(strconv.AppendInt(buf[:0], int64(a[0]), 10))

		for i := 1; err == nil && i < len(a); i++ {
			err = bw.WriteByte(opts.sep)
			if err == nil {
				_, err = bw.Write(strconv.AppendInt(buf[:0], int64(a[i]), 10))
			}
		}
	}

	if err == nil && opts.end != 0 {
		err = bw.WriteByte(opts.end)
	}

	return err
}

// ----------------------------------------------------------------------------

func gcd(a, b int) int {
	if a > b {
		a, b = b, a
	}
	for a > 0 {
		a, b = b%a, a
	}
	return b
}

func makeMatrix[T any](n, m int) [][]T {
	buf := make([]T, n*m)
	matrix := make([][]T, n)
	for i, j := 0, 0; i < n; i, j = i+1, j+m {
		matrix[i] = buf[j : j+m]
	}
	return matrix
}

File: p4+/main.py
def solve(t, s, v):
    count = 0

    v.sort()

    s0 = v[0] * t
    s0_i = 0
    s0_div_s = s0 // s
    s0_mod_s = s0 % s

    sum_div_s, si_count = 0, 0

    nodes = [Node() for _ in range(len(v))]
    tree = Node.insert(None, s0_mod_s, nodes)

    for i in range(1, len(v)):
        si = v[i] * t

        if si != s0:
            sum_div_s += s0_div_s * (i-s0_i)
            si_div_s = si // s
            si_count = (si_div_s-1)*i - sum_div_s
            s0 = si
            s0_i = i
            s0_div_s = si_div_s

        si_mod_s = si % s
        si_last = Node.find_idx(tree, si_mod_s)
        tree = Node.insert(tree, si_mod_s, nodes)
        s0_mod_s = si_mod_s

        count += si_count + si_last

    return count


class Node:
    def __init__(self):
        self._key = 0
        self._left = None
        self._right = None
        self._size = 1
        self._hght = 1

    @staticmethod
    def size(node):
        return 0 if node is None else node._size

    @staticmethod
    def height(node):
        return 0 if node is None else node._hght

    @staticmethod
    def find_idx(node, key):
        '''Возвращает индекс (начиная с 0) первого не меньшиго по ключу узла.'''
        '''Если такого ключа не существует, возвращает node.size()'''
        idx = 0
        while node is not None:
            if key <= node._key:
                node = node._left
            else:
                idx += Node.size(node._left) + 1
                node = node._right
        return idx

    @staticmethod
    def insert(node, key, nodes):
        if node is None:
            node = nodes.pop()
            node._key = key
            return node
        if key <= node._key:
            new_left = Node.insert(node._left, key, nodes)
            node._left = new_left
            return Node._repair(node)
        else:
            new_right = Node.insert(node._right, key, nodes)
            node._right = new_right
            return Node._repair(node)

    @staticmethod
    def _update(node):
        node._size = Node.size(node._left) + Node.size(node._right) + 1
        node._hght = max(Node.height(node._left), Node.height(node._right)) + 1

    @staticmethod
    def _repair(node):
        d = Node.height(node._left) - Node.height(node._right)
        if d < -1:
            return Node._left_rotate(node)
        elif d > 1:
            return Node._right_rotate(node)
        else:
            Node._update(node)
            return node

    @staticmethod
    def _left_rotate(node):
        al = node
        bt = al._right

        if Node.height(bt._right) - Node.height(bt._left) > 0:
            al._right = bt._left
            Node._update(al)
            bt._left = al
            Node._update(bt)
            return bt
        else:
            ga = bt._left
            al._right = ga._left
            Node._update(al)
            bt._left = ga._right
            Node._update(bt)
            ga._left = al
            ga._right = bt
            Node._update(ga)
            return ga

    @staticmethod
    def _right_rotate(node):
        al = node
        bt = al._left

        if Node.height(bt._left) - Node.height(bt._right) > 0:
            al._left = bt._right
            Node._update(al)
            bt._right = al
            Node._update(bt)
            return bt
        else:
            ga = bt._right
            al._left = ga._right
            Node._update(al)
            bt._right = ga._left
            Node._update(bt)
            ga._right = al
            ga._left = bt
            Node._update(ga)
            return ga


def main():
    n, t, s = map(int, input().split())
    v = list(map(int, input().split()))
    res = solve(t, s, v)
    print(res)


if __name__ == '__main__':
    main()
